core/src/Streamly/Internal/Data/MutByteArray/Type.hs

```diff
@@ -101,58 +102,46 @@ touch (MutByteArray contents) =
--- We can pass an unlifted pointer to the FFI routine to avoid GC kicking in
--- before the pointer is wrapped.
---
--- From the GHC manual:
---
--- GHC, since version 8.4, guarantees that garbage collection will never occur
--- during an unsafe call, even in the bytecode interpreter, and further
--- guarantees that unsafe calls will be performed in the calling thread. Making
--- it safe to pass heap-allocated objects to unsafe functions.
+-- | NOTE: this is deprecated because it can lead to accidental problems if the
+-- user tries to use it to mutate the array because it does not return the new
+-- array after pinning.
+{-# DEPRECATED unsafePinnedAsPtr "Pin the array and then use unsafeAsPtr." #-}
+{-# INLINE unsafePinnedAsPtr #-}
+unsafePinnedAsPtr :: MonadIO m => MutByteArray -> (Ptr a -> m b) -> m b
+unsafePinnedAsPtr arr f = do
+    arr1 <- liftIO $ pin arr
+    unsafeAsPtr arr1 f
+
+{-# DEPRECATED asPtrUnsafe "Pin the array and then use unsafeAsPtr." #-}
+{-# INLINE asPtrUnsafe #-}
+asPtrUnsafe :: MonadIO m => MutByteArray -> (Ptr a -> m b) -> m b
+asPtrUnsafe = unsafePinnedAsPtr
 
 -- | Use a @MutByteArray@ as @Ptr a@. This is useful when we want to pass
 -- an array as a pointer to some operating system call or to a "safe" FFI call.
 --
--- If the array is not pinned it is copied to pinned memory before passing it
--- to the monadic action.
+-- /Unsafe/ WARNING:
 --
--- /Performance Notes:/ Forces a copy if the array is not pinned. It is advised
--- that the programmer keeps this in mind and creates a pinned array
--- opportunistically before this operation occurs, to avoid the cost of a copy
--- if possible.
+-- 1. Will lead to memory corruption if the array is not pinned. Use
+-- only if the array is known to be pinned already or pin it explicitly.
 --
--- /Unsafe/ because of direct pointer operations. The user must ensure that
--- they are writing within the legal bounds of the array.
+-- 2. Ensure that the pointer is accessed within the legal bounds of the array.
+-- The size of the MutByteArray must be taken into account.
 --
 -- /Pre-release/
 --
-{-# INLINE unsafePinnedAsPtr #-}
-unsafePinnedAsPtr :: MonadIO m => MutByteArray -> (Ptr a -> m b) -> m b
-unsafePinnedAsPtr arr f = do
-  contents <- liftIO $ pin arr
-  let !ptr = Ptr (byteArrayContents#
-                     (unsafeCoerce# (getMutByteArray# contents)))
-  r <- f ptr
-  liftIO $ touch contents
-  return r
-
-{-# DEPRECATED asPtrUnsafe "Please use unsafePinnedAsPtr instead." #-}
-{-# INLINE asPtrUnsafe #-}
-asPtrUnsafe :: MonadIO m => MutByteArray -> (Ptr a -> m b) -> m b
-asPtrUnsafe = unsafePinnedAsPtr
-
--- | For use with unsafe FFI functions. Does not force pin the array memory.
 {-# INLINE unsafeAsPtr #-}
 unsafeAsPtr :: MonadIO m => MutByteArray -> (Ptr a -> m b) -> m b
 unsafeAsPtr arr f = do
-  let !ptr = Ptr (byteArrayContents#
+    -- XXX Make this an error instead?
+    assertM(isPinned arr)
```

Comment:
You can make this an error here.
I'm very sure, this will be misused.

Reply:
Why not pin it here, if unpinned?

Reply:
Oh, we shouldn't pin it. It'll be a different array then. This will cause programmer confusion.
Can error out though.